rosbag2_cpp/src/rosbag2_cpp/service_utils.cpp

```diff
@@ -0,0 +1,136 @@
+// Copyright 2023 Sony Group Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cstring>
+
+#include "rcl/service_introspection.h"
+
+#include "rosbag2_cpp/service_utils.hpp"
+#include "rosidl_typesupport_cpp/message_type_support_dispatch.hpp"
+#include "rosidl_typesupport_introspection_cpp/identifier.hpp"
+#include "rosidl_typesupport_introspection_cpp/message_introspection.hpp"
+
+#include "service_msgs/msg/service_event_info.hpp"
+
+namespace rosbag2_cpp
+{
+const char * service_event_topic_type_postfix = "_Event";
+const char * service_event_topic_type_middle = "/srv/";
+
+bool is_service_event_topic(const std::string & topic, const std::string & topic_type)
+{
+  if (topic.length() <= (sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1)) {
+    return false;
+  }
+
+  std::string end_topic_name = topic.substr(
+    topic.length() - (sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1));
+
+  if (topic_type.length() <= std::strlen(service_event_topic_type_postfix)) {
+    return false;
+  }
+
+  // Should include '/srv/' in type
+  if (topic_type.find(service_event_topic_type_middle) == std::string::npos) {
+    return false;
+  }
+
+  if (topic_type.length() <= std::strlen(service_event_topic_type_postfix)) {
+    return false;
+  }
+
+  return (topic_type.compare(
+           topic_type.length() - std::strlen(service_event_topic_type_postfix),
+           std::strlen(service_event_topic_type_postfix),
+           service_event_topic_type_postfix) == 0) &&
+         (end_topic_name == RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX);
+}
+
+std::string service_event_topic_name_to_service_name(const std::string & topic_name)
+{
+  if (topic_name.length() <= (sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1)) {
+    return std::string();
+  }
+
+  if (topic_name.substr(
+      topic_name.length() -
+      (sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1)) !=
+    RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX)
+  {
+    return std::string();
+  }
+
+  std::string service_name = topic_name.substr(
+    0, topic_name.length() - (sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1));
+
+  return service_name;
```

Comment:
clang-tidy giving a warning about: "Avoid repeating the return type from the declaration; use a braced initializer list instead" for the using `return std::string();`
But actually, the good practice is to declare the returning variable at the beginning and use it as the return value.
```suggestion
  std::string service_name;
  if (topic_name.length() <= (sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1)) {
    return service_name;
  }

  if (topic_name.substr(
      topic_name.length() - (sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1)) !=
    RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX)
  {
    return service_name;
  }

  service_name = topic_name.substr(
    0, topic_name.length() - (sizeof(RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX) - 1));

  return service_name;
```